### PR TITLE
Use ActiveSupport::MessageVerifier's expiry and purpose feature in SignedGlobalID

### PR DESCRIPTION
**Pull request by assain**

### CODE REVIEW

**Comment by assain** [comment]
@kaspth ActiveSupport::MessageVerfifier added support for expiry and purpose in Rails 5.2 so I updated the Gemfile.lock and fixed the SignedGlobalID tests and they work fine now. However, could you please help me connect the dots:

1. Should we add a config for backward compatibility since the signed value will be different from earlier?  (Travis is Failing).
2. I understand that GlobalId is used by models in Rails and is listed as dependencies for ActiveRecord and ActiveJob. Will this change get picked up by Rails through globalid version number?

Also, let me know if I am missing anything :smile:

**Comment by assain** [comment]
@kaspth If we're using a new Rails app then this works well. But considering the case of an old Rails app which uses gid's and upgrades to the latest Rails and Global ID version, shouldn't they still be parsable?

File: test/cases/railtie_test.rb [hunk 1]
```diff
@@ -38,7 +38,7 @@ def setup
 
   test 'SignedGlobalID.verifier defaults to nil when secret_token is not present' do
     @app.initialize!
-    assert_nil SignedGlobalID.verifier
+    assert_not_nil SignedGlobalID.verifier
   end
```
Comment (assain) [inline test/cases/railtie_test.rb]: secret_token is deprecated

**Comment by kaspth** [comment]
Yep, they should be parsable. We pretty much always want to keep backwardscompatibility.

File: test/cases/signed_global_id_test.rb [hunk 2]
```diff
@@ -78,17 +78,19 @@ def with_default_verifier(verifier)
 class SignedGlobalIDPurposeTest < ActiveSupport::TestCase
   setup do
     @login_sgid = SignedGlobalID.create(Person.new(5), for: 'login')
+    @like_sgid = SignedGlobalID.create(Person.new(5), for: 'like-button')
   end
 
   test 'sign with purpose when :for is provided' do
-    assert_equal "eyJnaWQiOiJnaWQ6Ly9iY3gvUGVyc29uLzUiLCJwdXJwb3NlIjoibG9naW4iLCJleHBpcmVzX2F0IjpudWxsfQ==--4b9630f3a1fb3d7d6584d95d4fac96433ec2deef", @login_sgid.to_s
+    assert_equal "eyJfcmFpbHMiOnsibWVzc2FnZSI6IkltZHBaRG92TDJKamVDOVFaWEp6YjI0dk5TST0iLCJleHAiOm51bGwsInB1ciI6ImxvZ2luIn19--c39de01a211a37d62b4773d1da7bff94ba2ec176", @login_sgid.to_s
+    assert_not_equal @login_sgid, @like_sgid
```
Comment (kaspth) [inline test/cases/signed_global_id_test.rb]: Let's just do:

```ruby
assert_not_equal @login_sgid, SignedGlobalID.create(Person.new(5), for: 'like-button')
```

File: test/cases/signed_global_id_test.rb [hunk 3]
```diff
@@ -227,7 +229,7 @@ class SignedGlobalIDCustomParamsTest < ActiveSupport::TestCase
   end
 
   test 'parse custom params' do
-    sgid = SignedGlobalID.parse('eyJnaWQiOiJnaWQ6Ly9iY3gvUGVyc29uLzU/aGVsbG89d29ybGQiLCJwdXJwb3NlIjoiZGVmYXVsdCIsImV4cGlyZXNfYXQiOm51bGx9--7c042f09483dec470fa1088b76d9fd946eb30ffa')
+    sgid = SignedGlobalID.parse('eyJfcmFpbHMiOnsibWVzc2FnZSI6IkltZHBaRG92TDJKamVDOVFaWEp6YjI0dk5UOW9aV3hzYnoxM2IzSnNaQ0k9IiwiZXhwIjpudWxsLCJwdXIiOiJkZWZhdWx0In19--09eb7f8baa6e0e19012dc4d1fd749e0c0cd33314')
     assert_equal 'world', sgid.params[:hello]
```
Comment (kaspth) [inline test/cases/signed_global_id_test.rb]: Would be a good idea to keep the old printed SGIDs around in separate tests to check that we can parse old gids.

File: lib/global_id/signed_global_id.rb [hunk 4]
```diff
@@ -30,6 +32,9 @@ def pick_purpose(options)
 
     private
       def verify(sgid, options)
+        gid = pick_verifier(options).verify(sgid, purpose: pick_purpose(options))
```
Comment (kaspth) [inline lib/global_id/signed_global_id.rb]: How about clarifying the flow and its purpose?

```ruby
verify_with_verifier_validated_metadata(sgid, options) || verify_with_self_validated_metadata(sgid, options)
```

**Comment by assain** [comment]
@kaspth Old sgids are now readable and new sgids are handled by Message Verifier. The tests have been updated. Also, Travis passes for 5.2.1.

However,  Travis is failing overall because -
1. 4.2 has a Minitest:: Runnable#marshal_dump is deprecated
2. 5.0 and 5.1 doesn't accept metadata as parameters in ActiveSupport:: MessageVerifier.

We also haven't added support for ruby 2.5. in Travis.

If everything's good, should we remove 4.2,5.0,5.2 from Travis?

Is there something I'm missing? :smile:

**Review by newblaque — APPROVED** [review]

**Comment by rafaelfranca** [comment]
Merged in 9d15bf2

**Pull request closed without merging**